src/System.Linq/tests/ConcatTests.cs

```diff
@@ -206,6 +206,16 @@ public void CountOfConcatIteratorShouldThrowExceptionOnIntegerOverflow()
             Assert.Throws<OverflowException>(() => supposedlyLargeCollection.Concat(tinyCollection).Count());
             Assert.Throws<OverflowException>(() => tinyCollection.Concat(tinyCollection).Concat(supposedlyLargeCollection).Count());
             Assert.Throws<OverflowException>(() => tinyCollection.Concat(tinyCollection).Concat(tinyCollection).Concat(supposedlyLargeCollection).Count());
+
+            // This applies to ToArray() and ToList() as well, which try to preallocate the exact size
+            // needed if all inputs are ICollections.
+            Assert.Throws<OverflowException>(() => supposedlyLargeCollection.Concat(tinyCollection).ToArray());
+            Assert.Throws<OverflowException>(() => tinyCollection.Concat(tinyCollection).Concat(supposedlyLargeCollection).ToArray());
+            Assert.Throws<OverflowException>(() => tinyCollection.Concat(tinyCollection).Concat(tinyCollection).Concat(supposedlyLargeCollection).ToArray());
+
+            Assert.Throws<OverflowException>(() => supposedlyLargeCollection.Concat(tinyCollection).ToList());
+            Assert.Throws<OverflowException>(() => tinyCollection.Concat(tinyCollection).Concat(supposedlyLargeCollection).ToList());
+            Assert.Throws<OverflowException>(() => tinyCollection.Concat(tinyCollection).Concat(tinyCollection).Concat(supposedlyLargeCollection).ToList());
```

Comment:
> At the very least we need to understand the exact circumstances that lead to it and make a conscious decision about whether to take the breaking change.

This will break only for badly behaving ICollections, e.g. for collections whose `Count` is greater than the number of items they actually have. If the sum of the collections' counts overflowed and the collections actually *did* have `> int.MaxValue` items, then obviously we would OOM somehow when trying to call `ToArray` previously. So for valid implementations, this is just changing the exception thrown from an OOME to an `OverflowException`, not throwing an exception on a path that previously worked.